### PR TITLE
[SPARK-27192][Core] spark.task.cpus should be less or equal than spark.executor.cores

### PR DESCRIPTION
**Pull request by liutang123**
## What changes were proposed in this pull request?
check spark.task.cpus before creating TaskScheduler in SparkContext

## How was this patch tested?
UT

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by liutang123** [comment]
@srowen @dongjoon-hyun 
Thanks for reviewing.
I solved some of the test failures.

**Comment by dongjoon-hyun** [comment]
Thank you, @liutang123 .

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 1]
```diff
@@ -155,7 +159,9 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
 
   test("Scheduler correctly accounts for multiple CPUs per task") {
     val taskCpus = 2
-    val taskScheduler = setupScheduler(config.CPUS_PER_TASK.key -> taskCpus.toString)
+    val taskScheduler = setupSchedulerWithMaster(
+      "local[2]",  // for passing CPUS_PER_TASK check in SparkContext
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: I don't know if that comment is necessary, and I suppose you could write `s"local[$taskCpus]"` here. But seems OK with me overall.

**Review by srowen — COMMENTED** [review]
@dongjoon-hyun as a double check this time does it look reasonable?

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -2665,8 +2665,27 @@ object SparkContext extends Logging {
     // When running locally, don't try to re-execute tasks on failure.
     val MAX_LOCAL_TASK_FAILURES = 1
 
+    // SPARK-26340: Ensure that executor's core num meets at least one task requirement.
+    def checkCpusPerTask(
+      clusterMode: Boolean,
+      maxCoresPerExecutor: Option[Int]): Unit = {
+      val cpusPerTask = sc.conf.get(CPUS_PER_TASK)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Missing `contains` check, here ?

Comment (liutang123) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: @Ngone51 `CPUS_PER_TASK` has default value 1, so, contains check is not needed.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 3]
```diff
@@ -577,16 +577,6 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
...
 
-    if (contains(EXECUTOR_CORES) && contains(CPUS_PER_TASK)) {
-      val executorCores = get(EXECUTOR_CORES)
-      val taskCpus = get(CPUS_PER_TASK)
-
-      if (executorCores < taskCpus) {
-        throw new SparkException(
-          s"${EXECUTOR_CORES.key} must not be less than ${CPUS_PER_TASK.key}.")
-      }
-    }
```
[the remark this answers is not in the document]
Comment (liutang123) [reply core/src/main/scala/org/apache/spark/SparkConf.scala]: You can see the difference at #24131 's discuss.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 4]
```diff
@@ -2665,8 +2665,27 @@ object SparkContext extends Logging {
...
+      if (clusterMode && sc.conf.contains(EXECUTOR_CORES)) {
+        if (sc.conf.get(EXECUTOR_CORES) < cpusPerTask) {
+          throw new SparkException(s"${CPUS_PER_TASK.key}" +
+            s" must be <= ${EXECUTOR_CORES.key} when run on $master.")
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: why is `$master` relevant here? I see we only check `clusterMode`

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 5]
```diff
@@ -2665,8 +2665,27 @@ object SparkContext extends Logging {
...
+        }
+      } else if (maxCoresPerExecutor.isDefined) {
+        if (maxCoresPerExecutor.get < cpusPerTask) {
+          throw new SparkException(s"Only ${maxCoresPerExecutor.get} cores available per executor" +
+            s" when run on $master, and ${CPUS_PER_TASK.key} must be <= it.")
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: ditto about `$master`

Comment (liutang123) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Max cores per executor is 3 when run on `local[3]` mode, and is 1 when run on `local` mode. I think we should info user that witch `master` he or she used.

**Review by srowen — APPROVED** [review]
@dongjoon-hyun @felixcheung I'll merge this tomorrow if you're OK

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**